### PR TITLE
feat: allow to use custom LLRT binary with `llrtBinaryPath`

### PR DESCRIPTION
**Pull request by georeeve**
Adds a `llrtCustomBinaryDirectory` property to allow users to use their own LLRT bootstrap binary.

This could be implemented in a few different ways. I tried to keep as much of the existing solution as possible:
If `llrtCustomBinaryDirectory` is set, it looks for a "bootstrap" binary in that directory, and errors if it cannot find one.
If `llrtCustomBinaryDirectory` is not set, the binary is downloaded from GitHub and cached in the `.tmp` directory as before.

### CODE REVIEW

**Comment by tmokmss** [comment]
@georeeve hi thanks for the PR! to clarify, what is the usecase for this feature? (Why do you want to configure the cache directory?) Thanks.

**Comment by tmokmss** [comment]
@georeeve Thanks for the detailed clarification! Actually I have some concerns about this proposed feature:

1. The bootstrap cache path becomes the same regardless of architecture, version, or binary type. It can confuse some users when they naively set llrtCustomBinaryDirectory prop.
2. This feature can prevent future changes in the caching strategy. I'd like to hide this part as implementation detail that users don't need to care.

Instead, I'd just overwrite `.tmp/llrt/**/bootstrap` file for testing purpose as you originally did. The path may change in the future indeed, but you can then adjust the binary path to overwrite.

**Comment by georeeve** [comment]
Those concerns are valid, I still think there's a use case for some projects to want to use a custom binary permanently in their project, perhaps if they apply some specific patches on top of LLRT. Would you accept a `llrtCustomBinary` property which specifies a file path to a specific bootstrap binary, and use this instead of altering the cache directory logic? We could then leave the caching strategy alone, and completely bypass the `afterBundling` step if `llrtCustomBinary` is provided.

**Comment by tmokmss** [comment]
I see. then let's add the feature! I'll take a close look at the PR tomorrow so just a moment please :pray:

File: src/llrt-function.ts [hunk 1]
```diff
@@ -123,6 +130,10 @@ export class LlrtFunction extends NodejsFunction {
           afterBundling: (i, o) => [
             // Download llrt binary from GitHub release and cache it
             `if [ ! -e ${posix.join(i, cacheDir, 'bootstrap')} ]; then
+              if [ -n "${props.llrtCustomBinaryDirectory}" ]; then
```
Comment (tmokmss) [inline src/llrt-function.ts:133]: doesnt this become `"undefined"`?

File: src/llrt-function.ts [hunk 2]
```diff
@@ -38,6 +38,13 @@ export interface LlrtFunctionProps extends NodejsFunctionProps {
    * @default LlrtBinaryType.STANDARD
    */
   readonly llrtBinaryType?: LlrtBinaryType;
+
+  /**
+   * A custom relative directory to find the LLRT "bootstrap" binary in.
+   *
+   * @default - The LLRT binary is downloaded from GitHub and cached in the .tmp directory.
+   */
+  readonly llrtCustomBinaryDirectory?: string;
```
Comment (tmokmss) [inline src/llrt-function.ts:47]: How about renaming it to `llrtBinaryPath` and take binary file path instead of directory path? Such name indicates more clearly that users have to explicitly prepare the binary.

File: src/llrt-function.ts [hunk 3]
```diff
@@ -123,6 +130,10 @@ export class LlrtFunction extends NodejsFunction {
           afterBundling: (i, o) => [
             // Download llrt binary from GitHub release and cache it
             `if [ ! -e ${posix.join(i, cacheDir, 'bootstrap')} ]; then
+              if [ -n "${props.llrtCustomBinaryDirectory}" ]; then
+                echo "Error: Could not find bootstrap binary in ${props.llrtCustomBinaryDirectory}"
+                exit 1
+              fi
               mkdir -p ${posix.join(i, cacheDir)}
               cd ${posix.join(i, cacheDir)}
               curl -L -o llrt_temp.zip ${binaryUrl}
```
Comment (tmokmss) [inline src/llrt-function.ts:139]: Let's skip the download part entirely when `props.llrtBinaryPath` is configured.

**Review by tmokmss — CHANGES_REQUESTED** [review]

File: src/llrt-function.ts (hunk 3 above)
Comment (georeeve) [reply src/llrt-function.ts:139]: Yeah, I think renaming it and changing the functionality is the best solution here

**Comment by georeeve** [comment]
@tmokmss Sorry for the delay, I've added the `llrtBinaryPath` prop now, let me know how it looks.

File: src/llrt-function.ts [hunk 4]
```diff
@@ -130,7 +137,7 @@ export class LlrtFunction extends NodejsFunction {
               rm -rf llrt_temp.zip
              fi`,
             `cp ${posix.join(i, cacheDir, 'bootstrap')} ${o}`,
-          ],
+          ] : [`cp ${posix.join(i, props.llrtBinaryPath)} ${o}`],
```
Comment (tmokmss) [inline src/llrt-function.ts:140]: Not sure if this works well in some edge cases. Does it work when`forceDockerBundling: true` and/or `depsLockFilePath: 'some/child/packege-lock.json'`? I'll test those cases later.

Comment (tmokmss) [reply src/llrt-function.ts:140]: Also we may need an integration test to maintain this feature. I'll add one along the above test.

**Review by tmokmss — COMMENTED** [review]
Thank you @georeeve! I'll confirm the detailed behavior in my local env tomorrow.  I think we can merge this after that :)

File: src/llrt-function.ts (hunk 4 above)
Comment (georeeve) [reply src/llrt-function.ts:140]: Good spot, I'm not sure if it would work with `forceDockerBundling`. I imagine you would have to mount the directory somehow but I don't know if there's an option for that?

Comment (georeeve) [reply src/llrt-function.ts:140]: @tmokmss I can take a look at the integration tests, does it just run with `yarn integ-runner`?

Comment (tmokmss) [reply src/llrt-function.ts:140]: yes. you may need to run yarn build for tsc though.

File: src/llrt-function.ts [hunk 5]
```diff
@@ -38,6 +38,13 @@ export interface LlrtFunctionProps extends NodejsFunctionProps {
    * @default LlrtBinaryType.STANDARD
    */
   readonly llrtBinaryType?: LlrtBinaryType;
+
+  /**
+   * A custom relative path to use as a local LLRT bootstrap binary.
```
[the remark this answers is not in the document]
Comment (georeeve) [reply src/llrt-function.ts:43]: I haven't, I can try but I'm not able to test the behaviour on Windows, which I imagine would be different. If Windows requires Docker builds then I suppose that's a different problem entirely though.

File: src/llrt-function.ts (hunk 4 above)
Comment (tmokmss) [reply src/llrt-function.ts:140]: Confirmed that as long as llrtBinaryPath is relative to [the `projectRoot` directory](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunction.html#projectroot), it works even with Docker. I added that information to jsdoc.

**Comment by tmokmss** [comment]
@georeeve I made some minor modification to the code and doc, as well as adding an integ-test. If you are okay with these changes, I'll merge this! :)

**Review by tmokmss — APPROVED** [review]

**Comment by georeeve** [comment]
@tmokmss LGTM, thanks for doing that.
